crates/walrus-upload-relay/src/controller.rs

```diff
@@ -0,0 +1,446 @@
+    pub(crate) n_shards: NonZeroU16,
+    pub(crate) metric_set: WalrusUploadRelayMetricSet,
+}
+
+impl Controller {
+    /// Creates a new controller.
+    pub(crate) fn new(
+        client: Client<SuiReadClient>,
+        n_shards: NonZeroU16,
+        relay_config: WalrusUploadRelayConfig,
+        metric_set: WalrusUploadRelayMetricSet,
+    ) -> Self {
+        Self {
+            client,
+            relay_config,
+            n_shards,
+            metric_set,
+        }
+    }
+
+    /// Checks the request and fans out the data to the storage nodes.
+    #[tracing::instrument(level = Level::DEBUG, skip_all)]
+    pub(crate) async fn fan_out(
+        &self,
+        body: Bytes,
+        params: Params,
+    ) -> Result<ResponseType, WalrusUploadRelayError> {
+        if self.relay_config.tip_config.requires_payment() {
+            // Check authentication pre-conditions for upload relay, if the configuration requires a
+            // tip.
+            let paid_params = params.to_paid_params()?;
+            self.validate_auth_package(&paid_params, body.as_ref())
+                .await?;
+        }
+
+        let encode_start_timer = Instant::now();
+        // PERF: encoding should probably be done on a separate thread pool.
+        let (sliver_pairs, metadata) = self
+            .client
+            .encoding_config()
+            .get_for_type(
+                params
+                    .encoding_type
+                    .unwrap_or(walrus_sdk::core::DEFAULT_ENCODING),
+            )
+            .encode_with_metadata(body.as_ref())?;
+        let duration = encode_start_timer.elapsed();
+
+        tracing::debug!(
+            computed_blob_id=%metadata.blob_id(),
+            expected_blob_id=%params.blob_id,
+            "blob id computed"
+        );
+
+        if *metadata.blob_id() != params.blob_id {
+            self.metric_set.blob_id_mismatch.inc();
+            return Err(WalrusUploadRelayError::BlobIdMismatch);
+        }
+
+        let pair = sliver_pairs
+            .first()
+            .expect("the encoding produces sliver pairs");
+        let symbol_size = pair.primary.symbols.symbol_size().get();
+
+        tracing::debug!(
+            symbol_size,
+            primary_sliver_size = pair.primary.symbols.data().len(),
+            secondary_sliver_size = pair.secondary.symbols.data().len(),
+            ?duration,
+            "encoded sliver pairs and metadata"
+        );
+
+        // Attempt to upload the slivers.
+        let blob_persistence = if let Some(object_id) = params.deletable_blob_object {
+            BlobPersistenceType::Deletable {
+                object_id: object_id.into(),
+            }
+        } else {
+            BlobPersistenceType::Permanent
+        };
+        let confirmation_certificate: ConfirmationCertificate = self
+            .client
+            .send_blob_data_and_get_certificate(&metadata, &sliver_pairs, &blob_persistence, None)
```

Comment:
Question: is there any functioning change in how the upload is done? Is it still that wait for 2/3 shards' response and wait for extra half of the previous time? As a service, if it can upload to more shards as it can (it can still return to the client right after 2/3 responses), this will going to help reducing blob recovery in storage node significantly!

Reply:
There is no functioning change to how upload is done. Yeah, once we get folks using this we should look into making it help with network data propagation, although, it's just like the client CLI, that would put more burden on the client egress costs, workload, memory load, etc. But, seems like it would be a good knob to add to configuration.